### PR TITLE
Backend request authz refinements

### PR DESCRIPTION
**Pull request by johakoch**
* make `OAuth2ReqAuth` and `TokenRequest` code more similar
* corrected error chaining
* don't propagate token request roundtrip errors
* add tests

---
<details>
    <summary>Reviewer checklist</summary>
    <ul>
        <li>Read PR description: a summary about the changes is required</li>
        <li>Changelog updated</li>
        <li>Documentation: docs/{Reference, Cli, ...}, Docker and cli help/usage</li>
        <li>Pulled branch, manually tested</li>
        <li>Verified requirements are met</li>
        <li>Reviewed the code</li>
        <li>Reviewed the related tests</li>
    </ul>
</details>

### CODE REVIEW

File: config/configload/validate.go [hunk 1]
```diff
@@ -35,10 +35,10 @@ func validLabel(name string, subject *hcl.Range) error {
 	return nil
 }
 
-func uniqueLabelName(unique map[string]struct{}, name string, hr *hcl.Range) error {
+func uniquePRLabelName(unique map[string]struct{}, name string, hr *hcl.Range) error {
 	if _, exist := unique[name]; exist {
 		if name == defaultNameLabel {
-			return newDiagErr(hr, "proxy and request labels are required and only one 'default' label is allowed")
+			return newDiagErr(hr, fmt.Sprintf("proxy and request labels are required and only one '%s' label is allowed", defaultNameLabel))
```
[the remark this answers is not in the document]
Comment (johakoch) [reply config/configload/validate.go:41]: only implicitly? what about
```hcl
proxy "default" {}
request "default" {}
```
?

[the remark this answers is not in the document]
Comment (alex-schneider) [reply config/configload/validate.go:41]: I mean the following:
```
endpoint "/" {
  proxy {}             # not required. it contradicts the "proxy and request labels are required"
  request "default" {} # required, but must not be "default"
}
```

[the remark this answers is not in the document]
Comment (johakoch) [reply config/configload/validate.go:41]: You're right, labels are not always _required_. But in your case, you'll probably get the error for the second block, which _explicitly_ has a `"default"` label.

[the remark this answers is not in the document]
Comment (johakoch) [reply config/configload/validate.go:41]: Would it be sufficient to have only one check here with message "proxy and request names (either default or explicitly set via label) must be unique"?

[the remark this answers is not in the document]
Comment (alex-schneider) [reply config/configload/validate.go:41]: good idea

File: config/configload/backend.go [hunk 2]
```diff
@@ -167,18 +167,61 @@ func wrapOAuthBackend(helper *helper, parent hcl.Body) (hcl.Body, error) {
...
 }
 
+func checkTokenRequestLabels(backendBody hcl.Body, unique map[string]struct{}) error {
+	ic, _, diags := backendBody.PartialContent(config.TokenRequestBlockSchema)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	trbs := ic.Blocks.OfType(tokenRequest)
+	if len(trbs) > 0 {
```
Comment (alex-schneider) [inline config/configload/backend.go:183]: This `if` is unnecessary.

File: config/configload/backend.go [hunk 3]
```diff
@@ -216,23 +255,20 @@ func wrapTokenRequestBackend(helper *helper, parent hcl.Body) (hcl.Body, error)
 			return nil, berr
 		}
 
+		// TODO what is this for? can it be removed?
```
Comment (alex-schneider) [inline config/configload/backend.go:258]: May be, the following `hclbody.MergeBodies()` would throw a panic

Comment (johakoch) [reply config/configload/backend.go:258]: `PrepareBackend()` only returns a `nil` tokenRequestBackend, if an error is returned, which is already handled by the preceding
```hcl
		if berr != nil {
			return nil, berr
		}
```

File: config/configload/backend.go (hunk 2 above)
Comment (johakoch) [reply config/configload/backend.go:183]: Done

File: config/configload/validate.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (johakoch) [reply config/configload/validate.go:41]: I simplified the name/label uniqueness checks.

**Review by alex-schneider — APPROVED** [review]

**Pull request merged**